rfcs/45-stageless.md

```diff
@@ -0,0 +1,641 @@
+
+There are two basic forms of system ordering constraints:
+
+1. **Strict ordering constraints:** `strictly_before` and `strictly_after`
+   1. A system cannot be scheduled until "strictly before" systems must have been completed during this iteration of the schedule.
+   2. Simple and explicit.
+   3. Can cause unneccessary blocking, particularly when systems are configured at a high-level.
+2. **If-needed ordering constraints:** `.before` and `.after`
+   1. A system cannot be scheduled until any "before" systems that it is incompatible with have completed during this iteration of the schedule.
+   2. In the vast majority of cases, this is the desired behavior. Unless you are using interior mutability, systems that are compatible will always be **commutative**: their ordering doesn't matter.
+
+Applying an ordering constraint to or from a label causes a ordering constraint to be created between all individual members of that label.
+If an ordering is defined relative to a non-existent system or an unused label, it will have no effect, emitting a warning.
+This relatively gentle failure mode is important to ensure that plugins can order their systems with relatively strong assumptions that the default system labels exist, but continue to (mostly) work if those systems or labels are not present.
+
+In addition to the `.before` and `.after` methods, you can use **system chains** to create very simple linear dependencies between the succesive members of an array of systems.
+(Note to readers: this is not the same as "system chaining" in Bevy 0.6 and earlier: that concept has been renamed to "system handling".)
+
+When discussing system ordering, it is particularly important to call out the `flush_commands` system.
+This **exclusive system** (meaning, it can modify the entire `World` in arbitrary ways and cannot be run in parallel with other systems) collects all created commands and applies them to the `World`.
+
+Commands (commonly used to spawn and despawn entities or add and remove components) will not take effect until this system is run, so be sure that you run a copy of the `flush_commands` system before relying on the result of a command!
+
+This pattern is so common that a special form of ordering constraint exists for it: **command-flushed ordering constraints**.
+If system `A` is `before_and_flush` system `B`, the schedule will be unsatisfiable unless there is an intervening `flush_commands` system.
+Note that **this does not insert new copies of a `flush_commands` system**: instead, it functions like an `assert!` statement.
+It has no direct effect, and is merely used to verify that your schedule has been set up correctly according the specified constraint.
+
+### Run criteria
+
+While ordering constraints determine *when* a system will run, **run criteria** will determine *if* it will run at all.
+A run criterion is a special kind of system which can read (but not write) data from the `World` and returns a boolean value.

```

Comment:
Differentiate this. Systems can be coerced into run criteria.

This is important to allow us to special-case and optimize simple checks.